File: front/src/lib/nocodb.ts
import { TABLES } from "./config";
import type { NocoRecord } from "./types";

export class NocoDbError extends Error {
  constructor(
    message: string,
    public status: number,
  ) {
    super(message);
    this.name = "NocoDbError";
  }
}

export interface ListRecordsOptions {
  where?: string;
  fields?: string[];
  dateFrom?: { field: string; iso: string };
  dateTo?: { field: string; iso: string };
  limit?: number;
  offset?: number;
  sort?: string;
}

interface RecordsPage {
  list: NocoRecord[];
  pageInfo: { isLastPage: boolean; totalRows: number };
}

function buildWhere(options: ListRecordsOptions): string | undefined {
  const parts: string[] = [];
  if (options.where) parts.push(options.where);
  if (options.dateFrom) {
    parts.push(`(${options.dateFrom.field},ge,exactDate,${options.dateFrom.iso})`);
  }
  if (options.dateTo) {
    parts.push(`(${options.dateTo.field},le,exactDate,${options.dateTo.iso})`);
  }
  if (parts.length === 0) return undefined;
  if (parts.length === 1) return parts[0];
  return parts.join("~and");
}

function normalizeListOptions(
  arg?: string | ListRecordsOptions,
  legacyFields?: string[],
): ListRecordsOptions {
  if (typeof arg === "string") {
    return { where: arg, fields: legacyFields };
  }
  return arg ?? {};
}

export class NocoDbClient {
  constructor(
    private baseUrl: string,
    private token: string,
  ) {}

  private async request<T>(path: string, init?: RequestInit): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      ...init,
      headers: {
        "Content-Type": "application/json",
        "xc-token": this.token,
        ...init?.headers,
      },
      cache: "no-store",
    });

    if (!res.ok) {
      const body = await res.text();
      throw new NocoDbError(body || res.statusText, res.status);
    }

    return res.json() as Promise<T>;
  }

  private async fetchRecordsPage(
    tableId: string,
    options: ListRecordsOptions,
  ): Promise<RecordsPage> {
    const pageSize = options.limit ?? 500;
    const offset = options.offset ?? 0;
    const where = buildWhere(options);

    const params = new URLSearchParams({
      limit: String(pageSize),
      offset: String(offset),
    });
    if (where) params.set("where", where);
    if (options.fields?.length) params.set("fields", options.fields.join(","));
    if (options.sort) params.set("sort", options.sort);

    return this.request<RecordsPage>(`/api/v2/tables/${tableId}/records?${params}`);
  }

  async countRecords(tableId: string, where?: string): Promise<number> {
    const data = await this.fetchRecordsPage(tableId, {
      where,
      fields: ["Id"],
      limit: 1,
    });
    return data.pageInfo.totalRows;
  }

  /** Años disponibles con dos consultas (min/max fecha), sin cargar toda la tabla. */
  async availableYears(
    tableId: string,
    dateField: string,
    where?: string,
  ): Promise<{ min: string | null; max: string | null }> {
    const base = { where, fields: [dateField], limit: 1 };
    const [oldest, newest] = await Promise.all([
      this.fetchRecordsPage(tableId, { ...base, sort: dateField }),
      this.fetchRecordsPage(tableId, { ...base, sort: `-${dateField}` }),
    ]);

    const min = oldest.list[0]?.[dateField];
    const max = newest.list[0]?.[dateField];
    return {
      min: min ? String(min).slice(0, 10) : null,
      max: max ? String(max).slice(0, 10) : null,
    };
  }

  async listRecords(
    tableId: string,
    arg?: string | ListRecordsOptions,
    legacyFields?: string[],
  ): Promise<NocoRecord[]> {
    const options = normalizeListOptions(arg, legacyFields);

    if (options.limit !== undefined || options.offset !== undefined) {
      const page = await this.fetchRecordsPage(tableId, options);
      return page.list;
    }

    const pageSize = 500;
    const first = await this.fetchRecordsPage(tableId, { ...options, limit: pageSize, offset: 0 });
    const records = [...first.list];

    if (first.pageInfo.isLastPage || first.list.length === 0) {
      return records;
    }

    const total = first.pageInfo.totalRows;
    const offsets: number[] = [];
    for (let offset = pageSize; offset < total; offset += pageSize) {
      offsets.push(offset);
    }

    const pages = await Promise.all(
      offsets.map((offset) =>
        this.fetchRecordsPage(tableId, { ...options, limit: pageSize, offset }),
      ),
    );

    for (const page of pages) {
      records.push(...page.list);
    }

    return records;
  }

  async findUserByUsername(username: string): Promise<NocoRecord | null> {
    const normalized = username.trim().toLowerCase();
    const data = await this.request<{ list: NocoRecord[] }>(
      `/api/v2/tables/${TABLES.users}/records?limit=100`,
    );
    return (
      data.list.find((row) => String(row.Username ?? "").toLowerCase() === normalized) ?? null
    );
  }

  async getTableMeta(tableId: string): Promise<{ columns?: { title: string; colOptions?: { options?: { title?: string }[] } }[] }> {
    return this.request(`/api/v2/meta/tables/${tableId}`);
  }

  async getRecord(tableId: string, id: string): Promise<NocoRecord | null> {
    const data = await this.request<{ list: NocoRecord[] }>(
      `/api/v2/tables/${tableId}/records?where=(Id,eq,${id})&limit=1`,
    );
    return data.list[0] ?? null;
  }

  async createRecord(tableId: string, fields: NocoRecord): Promise<NocoRecord> {
    return this.request(`/api/v2/tables/${tableId}/records`, {
      method: "POST",
      body: JSON.stringify(fields),
    });
  }

  async createRecords(tableId: string, records: NocoRecord[]): Promise<NocoRecord[]> {
    if (records.length === 0) return [];
    const data = await this.request<NocoRecord | NocoRecord[]>(
      `/api/v2/tables/${tableId}/records`,
      {
        method: "POST",
        body: JSON.stringify(records),
      },
    );
    return Array.isArray(data) ? data : [data];
  }

  async updateRecord(tableId: string, id: string, fields: NocoRecord): Promise<void> {
    await this.request(`/api/v2/tables/${tableId}/records`, {
      method: "PATCH",
      body: JSON.stringify({ Id: id, ...fields }),
    });
  }

  async deleteRecord(tableId: string, id: string): Promise<void> {
    await this.request(`/api/v2/tables/${tableId}/records`, {
      method: "DELETE",
      body: JSON.stringify([{ Id: id }]),
    });
  }
}

export function yearsFromDateBounds(
  minIso: string | null,
  maxIso: string | null,
  timezone: string,
  fallbackYear: string,
): string[] {
  if (!minIso || !maxIso) return [fallbackYear];

  const minDate = new Date(`${minIso.slice(0, 10)}T12:00:00`);
  const maxDate = new Date(`${maxIso.slice(0, 10)}T12:00:00`);
  const minYear = Number(
    new Intl.DateTimeFormat("en-CA", { timeZone: timezone, year: "numeric" }).format(minDate),
  );
  const maxYear = Number(
    new Intl.DateTimeFormat("en-CA", { timeZone: timezone, year: "numeric" }).format(maxDate),
  );

  const years: string[] = [];
  for (let y = minYear; y <= maxYear; y += 1) {
    years.push(String(y));
  }
  return years.length > 0 ? years : [fallbackYear];
}

File: worker/src/worker/nocodb.py
from __future__ import annotations

import urllib.parse
from typing import Any

import httpx


class NocoDbError(Exception):
    def __init__(self, message: str, status: int) -> None:
        super().__init__(message)
        self.status = status


class NocoDbClient:
    def __init__(self, base_url: str, token: str) -> None:
        self.base_url = base_url.rstrip("/")
        self.token = token

    async def list_records(
        self,
        table_id: str,
        *,
        where: str | None = None,
        fields: list[str] | None = None,
        limit: int = 1000,
        offset: int = 0,
    ) -> list[dict[str, Any]]:
        params: dict[str, str] = {
            "limit": str(limit),
            "offset": str(offset),
        }
        if where:
            params["where"] = where
        if fields:
            params["fields"] = ",".join(fields)

        data = await self._request(
            "GET",
            f"/api/v2/tables/{table_id}/records?{urllib.parse.urlencode(params)}",
        )
        return list(data.get("list", []))

    async def create_record(self, table_id: str, fields: dict[str, Any]) -> dict[str, Any]:
        return await self._request("POST", f"/api/v2/tables/{table_id}/records", json=fields)

    async def update_record(self, table_id: str, record_id: int | str, fields: dict[str, Any]) -> None:
        await self._request(
            "PATCH",
            f"/api/v2/tables/{table_id}/records",
            json={"Id": record_id, **fields},
        )

    async def delete_records(self, table_id: str, record_ids: list[int | str]) -> None:
        if not record_ids:
            return
        payload = [{"Id": record_id} for record_id in record_ids]
        await self._request("DELETE", f"/api/v2/tables/{table_id}/records", json=payload)

    async def create_records(self, table_id: str, records: list[dict[str, Any]]) -> list[dict[str, Any]]:
        if not records:
            return []
        data = await self._request("POST", f"/api/v2/tables/{table_id}/records", json=records)
        if isinstance(data, list):
            return data
        return [data]

    async def count_records(self, table_id: str) -> int:
        data = await self._request(
            "GET",
            f"/api/v2/tables/{table_id}/records?limit=1&fields=Id",
        )
        return int(data.get("pageInfo", {}).get("totalRows", 0))

    async def get_table_meta(self, table_id: str) -> dict[str, Any]:
        return await self._request("GET", f"/api/v2/meta/tables/{table_id}")

    async def list_base_tables(self, base_id: str) -> list[dict[str, Any]]:
        data = await self._request("GET", f"/api/v2/meta/bases/{base_id}/tables")
        return list(data.get("list", []))

    async def patch_column(self, column_id: str, body: dict[str, Any]) -> dict[str, Any]:
        return await self._request("PATCH", f"/api/v2/meta/columns/{column_id}", json=body)

    async def create_column(self, table_id: str, body: dict[str, Any]) -> dict[str, Any]:
        return await self._request("POST", f"/api/v2/meta/tables/{table_id}/columns", json=body)

    async def existing_idempotency_keys(
        self,
        table_id: str,
        keys: list[str],
        *,
        chunk_size: int = 50,
    ) -> set[str]:
        if not keys:
            return set()

        found: set[str] = set()
        for index in range(0, len(keys), chunk_size):
            chunk = keys[index : index + chunk_size]
            quoted = ",".join(chunk)
            where = f"(IdempotencyKey,in,{quoted})"
            rows = await self.list_records(table_id, where=where, fields=["IdempotencyKey"])
            for row in rows:
                key = row.get("IdempotencyKey")
                if key:
                    found.add(str(key))
        return found

    async def _request(self, method: str, path: str, **kwargs: Any) -> dict[str, Any]:
        headers = {"xc-token": self.token, "Content-Type": "application/json"}
        async with httpx.AsyncClient(base_url=self.base_url, timeout=60.0) as client:
            response = await client.request(method, path, headers=headers, **kwargs)
        if response.status_code >= 400:
            raise NocoDbError(response.text or response.reason_phrase, response.status_code)
        if not response.content:
            return {}
        return response.json()
